### ISSUE TITLE
Broken pull requests merge action after cherry picking

### ISSUE THREAD

**Issue by WuShell** [gogs/gogs#6963]
### Gogs version

0.12.3 and 0.12.7

### Git version

- Server: 2.36.1
- Client: -

### Operating system

FreeBSD 12.2-stable

### Database

PostgreSQL 11.16

### Describe the bug

Clicking on the "Merge" button on a PR crashes gogs, needing a full restart of the service to get it up again.

### To reproduce

I've a repo with this structure:

- master
    - dev
        - feature/implement-whatever
        - fix/incorrect-handling-whatever
        - etc
    - release/1.0.0
    - release/1.1.0
    - release/1.1.1
    - etc

The workflow is, working on dev until we get to a point where we can make a release, create a branch release/1.1.1 (for example) and then merge that into master (using gogs, through a PR). Release branches are kept in the repo, but the feature/fix/etc branches are deleted after merged into dev.

So far so good, worked pretty well.

Now, with dev running 1.2.0 we needed to "backport" some fixes made in the dev branch, into a branch release/1.1.2 (bugfix for 1.1.x) without pulling all the work done in dev for the next release.

We created release/1.1.2 (from release/1.1.1) and then used cherry-pick to backport those fixes from dev. We did double check the branch locally, no conflicts, all ok and we pushed it to the main repo.

After that, we kept working normally and on the next merge of a fix/feature branch into dev, through gogs PR interface, gogs blows up and stops working (restart is needed).

In the log files we can see this message:

```
[Macaron] PANIC: Value not found for type *repo.HTTPContext
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/context.go:123 (0x1b62934)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/context.go:112 (0x1b75a65)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/recovery.go:161 (0x1b75a58)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/logger.go:40 (0x1b66377)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/github.com/go-macaron/inject/inject.go:157 (0x1b2fa54)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/github.com/go-macaron/inject/inject.go:135 (0x1b2f7f9)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/context.go:121 (0x1b627bc)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/router.go:187 (0x1b76cb0)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/router.go:303 (0x1b6ede5)
/wrkdirs/usr/ports/devel/gogs/work/gogs-0.12.3/vendor/gopkg.in/macaron.v1/macaron.go:220 (0x1b676ed)
/usr/local/go/src/net/http/server.go:2843 (0x1aca622)
/usr/local/go/src/net/http/server.go:1925 (0x1ac5e2c)
/usr/local/go/src/runtime/asm_amd64.s:1374 (0x1774060)
```

(that's the first appearance of the error in our log files)

We did upgrade to 0.12.7, gogs still crashes, this is the output in the logs:

```
[Macaron] PANIC: Value not found for type *repo.HTTPContext
gopkg.in/macaron.v1@v1.3.9/context.go:123 (0x1a016bd)
gopkg.in/macaron.v1@v1.3.9/context.go:112 (0x1a07346)
gopkg.in/macaron.v1@v1.3.9/recovery.go:161 (0x1a0733d)
gopkg.in/macaron.v1@v1.3.9/logger.go:40 (0x1a048e4)
github.com/go-macaron/inject@v0.0.0-20160627170012-d8a0b8677191/inject.go:157 (0x19d8714)
github.com/go-macaron/inject@v0.0.0-20160627170012-d8a0b8677191/inject.go:135 (0x19d863c)
gopkg.in/macaron.v1@v1.3.9/context.go:121 (0x1a015d2)
gopkg.in/macaron.v1@v1.3.9/router.go:187 (0x1a0d1c4)
gopkg.in/macaron.v1@v1.3.9/router.go:303 (0x1a0e32c)
gopkg.in/macaron.v1@v1.3.9/macaron.go:220 (0x1a06464)
net/http/server.go:2916 (0x198141a)
net/http/server.go:1966 (0x197c8d6)
runtime/asm_amd64.s:1571 (0x1671f80)
```

### Expected behavior

I'd expect to get PRs merged when I click on the "Merge" button in a PR page

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by unknwon** [comment]
Thanks for the feedback!

There is [only one place asks for `*repo.HTTPContext`](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/gogs/gogs%24%40release/0.12+file:%5Einternal/route/repo/http%5C.go+*HTTPContext&patternType=literal) which is the HTTP push/pull endpoint, seems like the URL somehow gets mismatched.

If possible, could you share the URL path for the crash URL? My hunch is that `/` isn't getting correct URL encoded in your branch name from the web UI (i.e. `release/1.1.2` should be `release%2F1.1.2` as the form's "action").

**Comment by WuShell** [comment]
This is the extract from the web server log file for a POST request (merging a PR) that caused the crash:

(I've replaced some sensitive data like ip addresses or full paths)
```
::ffff:XX.XX.XX.XX - - [24/May/2022:11:34:28 +0200] "POST /path/to/repo/pulls/877/merge HTTP/1.1" 302 0 "-" "Mozilla/5.0 (X11; FreeBSD amd64; rv:95.0) Gecko/20100101 Firefox/95.0" "-"
::ffff:XX.XX.XX.XX - - [24/May/2022:11:34:28 +0200] "GET /path/to/repo/pulls/877 HTTP/1.1" 502 157 "-" "Mozilla/5.0 (X11; FreeBSD amd64; rv:95.0) Gecko/20100101 Firefox/95.0" "-"
```

That 502 there is the gateway timeout the web server returns as it is not able to connect to the gogs backend (it crashed just after the merge, while the merge has been done successfully)

After a restart (and log in again) I'm able to use gogs normally, until I do another PR merge (this has happened to every PR merge since we do the cherry-pick of those commits).

Hope this helps, let me know If I can provide any more info.

**Comment by unknwon** [comment]
I couldn't reproduce through cherry-picking, but I did managed to reproduce the same panic case while don't understand why that would happen after merge ... very a stupid error though 😂

**Comment by unknwon** [comment]
If I try to send a POST to the PR URL, e.g. `http://localhost:3000/path/to/repo/pulls/877** instead of a GET would hit this code path, maybe something somehow did this but it appear in your web server log though ...

**Comment by WuShell** [comment]
But, you have a point in your previous comment. Our branch names all end in the issue number, like:

*fix/manage_max_slots_per_year_zero_#876*

That one, for example, was merged to the "dev" branch.

So, maybe some part of the code handling the merge blows up because of */* or *#*?

The strange part is that it has been working for a long time without problems, it started happening after the cherry-pick thing.

**Comment by WuShell** [comment]
>
> If I try to send a POST to the PR URL, e.g. http://localhost:3000/path/to/repo/pulls/877 instead of a GET would hit this code path, maybe something somehow did this but it should appear in your web server log though ...
>

I'll try to do the next merge directly in the gogs server, without passing through the web server, see if I can reproduce it there too.

**Comment by unknwon** [comment]
While I can fix the panic for now, I'm afraid that something else is acting up to hit this code path 🤔

**Comment by WuShell** [comment]
>
> It is a [one line fix](https://github.com/gogs/gogs/pull/7003/files) so if any chance you could patch it before a formal patch release, it would provide us what's actual URL path is hitting it.
>

I'm using the freebsd packaged gogs in this server, but I'll try grabbing from git and building it.

**Comment by WuShell** [comment]
Ok, built gogs from sources (branch *jc/http-panic*), go 1.18.2, git 2.36.1, then started it instead of our previous version.

Then I created a new branch in our repo, from dev, called test/new-gogs-fix, and modified a README file. I pushed those changes to the new branch, then in gogs I created a new PR to merge test/new-gogs-fix into dev, all good. Then I clicked merge and got a 502 gateway timeout, gogs crashed.

This is what I see in the logs now:

```
2022/06/03 18:01:36 [TRACE] Session ID: 3bbd1715c9dc3f3d
2022/06/03 18:01:36 [TRACE] CSRF Token: TXcV2RAeLpoG3aRJa1895ub-7DI6MTY1NDI3MTkxODQ3Mjg0NDM1Mw
2022/06/03 18:01:36 [TRACE] NewMessageFrom (htmlBody):
<!DOCTYPE html>
<html>
<head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <title>[unified_portal] test gogs fix for our PR merge crashes (#881)</title>
</head>

<body>
        <p><p>This pr is merged on a new version of gogs, as a POC</p>
</p>
        <p>
                ---
                <br>
                <a href="https://whatever.com/path/to/repo/pulls/881">View it on Gogs</a>.
        </p>
</body>
</html>
2022/06/03 18:01:38 [TRACE] Pull request merged: 445
2022/06/03 18:01:38 [TRACE] DeliverHooks [repo_id: 1]
2022/06/03 18:01:38 [TRACE] AddTestPullRequestTask [head_repo_id: 1, head_branch: dev]: finding pull requests
2022/06/03 18:01:38 [TRACE] addHeadRepoTasks[439]: composing new test task
2022/06/03 18:01:38 [TRACE] PullRequest[439].UpdatePatch: ignored corrupted data
2022/06/03 18:01:38 [TRACE] PushToBaseRepo[1]: pushing commits to base repo 'refs/pull/873/head'
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x296b402]

goroutine 865 [running]:
gogs.io/gogs/internal/db.(*Repository).mustOwner(0x1780208?, {0x17897f0?, 0xc0000a2000?})
        /home/wu/repos/gogs/internal/db/repo.go:411 +0x22
gogs.io/gogs/internal/db.(*Repository).repoPath(0x0, {0x17897f0?, 0xc0000a2000?})
        /home/wu/repos/gogs/internal/db/repo.go:536 +0x28
gogs.io/gogs/internal/db.(*Repository).RepoPath(...)
        /home/wu/repos/gogs/internal/db/repo.go:540
gogs.io/gogs/internal/db.(*PullRequest).PushToBaseRepo(0xc0028d8460)
        /home/wu/repos/gogs/internal/db/pull.go:674 +0xfa
gogs.io/gogs/internal/db.addHeadRepoTasks({0xc00226db90, 0x1, 0x3?})
        /home/wu/repos/gogs/internal/db/pull.go:766 +0x125
gogs.io/gogs/internal/db.AddTestPullRequestTask(0xc001d8a200, 0x0?, {0xc002ecd250, 0x3}, 0x0)
        /home/wu/repos/gogs/internal/db/pull.go:811 +0x290
created by gogs.io/gogs/internal/db.(*PullRequest).Merge.func1
        /home/wu/repos/gogs/internal/db/pull.go:200 +0x14b
```

Hope it helps :sweat_smile:

**Comment by unknwon** [comment]
Thanks for the help! Really appreciated.

This log line:

```
2022/06/03 18:01:38 [TRACE] PullRequest[439].UpdatePatch: ignored corrupted data
```

Indicates that one of the PR has lost its originated repository (e.g. the fork repository has been removed). Again ... not sure why that would happen if all PRs are coming from the same repository.

Made another fix in https://github.com/gogs/gogs/pull/7004 to the new panic discovered 🤦‍♂️

**Comment by WuShell** [comment]
Thanks a lot for the time you put on this.
I'll give it a try with that new fix, see what happens in our env here.

>
> Indicates that one of the PR has lost its originated repository (e.g. the fork repository has been removed). Again ... not sure why that would happen if all PRs are coming from the same repository.
>

Meanwhile,` thinking about what you wrote there, and if the following would explain why it is happening to us:

We have one repo, master branch and a dev branch, with feature/whatever or fix/whatever, etc branches where developers work. 

When a feature/fix/etc is ready to review, we make a PR of that branch to dev and review starts, once it has been reviewed, it is merged to dev. 

We keep on like this until all features/fixes for a given release are done, then we create branch release/X.X.X from dev at that point. More review and testing go into that release branch and, once all is ready, we merge release/X.X.X into master and into dev. Then we move on to the next release cycle.

So far, so good.

Now, enter the cherry pick. 

As explained in the issue, we needed to put some fixes from the current dev branch into the codebase on release/1.1.3. 

So, I did create a branch of release/1.1.3 into release/1.1.4, cherry picked the commits from dev into release/1.1.4 and I made a PR in gogs from release/1.1.4 into release/1.1.3 (so the usual review/test procedure could be done, checking which changes are new since release/1.1.3). 

**But that PR was closed without merge** (we didn't want to merge those branches reallly) **and the branch was not merged back into master and dev, as usual.** 

Could that be the cause of the error? Maybe that unmerged PR release/1.1.4 into release/1.1.3 (even if it is closed) or the branch not being merged into the master/dev branch in the repo? 

Is there any way we can delete completely that unmerged PR release/1.1.4 into release/1.1.3? (I guess we could simply delete it from the database directly?

**Comment by WuShell** [comment]
Built gogs from sources (main branch) and tried again, this time the PR merge did not crash.

Thanks a lot @unknwon for fixing this.

**Comment by unknwon** [comment]
> **But that PR was closed without merge** (we didn't want to merge those branches reallly) **and the branch was not merged back into master and dev, as usual.**
> 
> Could that be the cause of the error? Maybe that unmerged PR release/1.1.4 into release/1.1.3 (even if it is closed) or the branch not being merged into the master/dev branch in the repo?

Thanks for the clue!

However, PRs closed but not merged should be _valid_, I'll do a quick check on this see if anything stands out 😅

**Comment by unknwon** [comment]
Unless there is a bug in the ORM (which most of the time not), I'd call it a day 😶‍🌫️ 

https://github.com/gogs/gogs/blob/a328e7ccc4f94e19cddfe5894636228663f5c7fa/internal/db/pull.go#L543-L544
